File: scripts/day4-part1-optimized.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>

using namespace std;

int solve() {
    ifstream file("../inputs/day-4-input.txt");
    vector<string> grid;
    string line;
    
    while (getline(file, line)) {
        grid.push_back(line);
    }
    file.close();
    
    int rows = grid.size();
    int cols = grid[0].size();
    int result = 0;
    
    // 8 directions
    static const int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    static const int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};
    
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (grid[i][j] == '@') {
                int count = 0;
                
                // Inline neighbor counting for better performance
                for (int d = 0; d < 8; d++) {
                    int newRow = i + dx[d];
                    int newCol = j + dy[d];
                    if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && grid[newRow][newCol] == '@') {
                        count++;
                        if (count >= 4) break; // Early exit optimization
                    }
                }
                
                if (count < 4) {
                    result++;
                }
            }
        }
    }
    
    return result;
}

int main() {
    const int iterations = 1;
    int answer = 0;
    
    auto start = chrono::high_resolution_clock::now();
    
    for (int i = 0; i < iterations; i++) {
        answer = solve();
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::microseconds>(end - start);
    
    double avgTime = static_cast<double>(duration.count()) / iterations;
    double totalTimeMs = duration.count() / 1000.0;
    
    cout << "C++ (Day 4 Part 1 - Optimized)" << endl;
    cout << "Answer: " << answer << endl;
    cout << "Average time: " << avgTime << " microseconds" << endl;
    cout << "Total time: " << totalTimeMs << " ms" << endl;
    
    return 0;
}

File: scripts/day4-part1-optimized.py
import time

def solve(grid):
    rows = len(grid)
    cols = len(grid[0])
    accessible_count = 0
    
    directions = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1), (0, 1),
        (1, -1), (1, 0), (1, 1)
    ]
    
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '@':
                adjacent_rolls = 0
                
                for dr, dc in directions:
                    nr, nc = r + dr, c + dc
                    
                    if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '@':
                        adjacent_rolls += 1
                
                if adjacent_rolls < 4:
                    accessible_count += 1
    
    return accessible_count

# Read input
with open("../inputs/day-4-input.txt", "r") as f:
    input_data = f.read().strip().split('\n')

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 4 Part 1 - Optimized)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")
